File: Angular/src/app/auth.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Router } from '@angular/router';
import { BehaviorSubject, Observable, throwError } from 'rxjs';
import { tap, map, catchError } from 'rxjs/operators';

@Injectable({
  providedIn: 'root'
})
export class AuthService {
  private apiUrl = 'http://localhost:8000/api';
  private sessionExpired$ = new BehaviorSubject<boolean>(false);

  constructor(private http: HttpClient, private router: Router) {}

  // Register a new user
  register(data: any): Observable<any> {
    return this.http.post(`${this.apiUrl}/register`, data);
  }

  // Log in an existing user
  login(data: any): Observable<{ token: string }> {
    return this.http.post<{ token: string }>(`${this.apiUrl}/login`, data).pipe(
      tap(response => {
        if (response.token) {
          this.setToken(response.token);
        }
      })
    );
  }

  // Delete user account
  deleteAccount(password: string): Observable<any> {
    return this.http.post(`${this.apiUrl}/delete-account`, { password }, {
      headers: {
        Authorization: `Bearer ${this.getToken()}`
      }
    }).pipe(
      tap(() => {
        this.clearToken();
      }),
      catchError((error) => {
        return throwError(() => error);
      })
    );
  }

  // Update profile (name and email)
  updateProfile(data: { name: string; email: string }): Observable<any> {
    return this.http.put(`${this.apiUrl}/update-profile`, data, {
        headers: {
            Authorization: `Bearer ${this.getToken()}`
        }
    });
  }

  // Get user profile
  getProfile(): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/profile`, {
      headers: {
        Authorization: `Bearer ${this.getToken()}`
      }
    });
  }

  // Update password
  updatePassword(data: { current_password: string; new_password: string }): Observable<any> {
    return this.http.put(`${this.apiUrl}/update-password`, data, {
        headers: {
            Authorization: `Bearer ${this.getToken()}`
        }
    });
  }

  // Store the token in local storage
  setToken(token: string) {
    localStorage.setItem('token', token);
  }

  // Get the token from local storage
  getToken(): string | null {
    return localStorage.getItem('token');
  }

  // Remove the token from local storage
  clearToken() {
    localStorage.removeItem('token');
  }

  getSessionExpired(): Observable<boolean> {
    return this.sessionExpired$.asObservable();
  }

  // Check if the token is expired
  checkTokenExpiry(): Observable<boolean> {
    const token = this.getToken();
    if (!token) {
      this.clearToken();
      return new BehaviorSubject(false).asObservable();
    }

    return this.http.post<{ expired: boolean }>(`${this.apiUrl}/check-token-expiry`, { token }).pipe(
      map(response => {
        if (response.expired) {
          this.clearToken();
          this.sessionExpired$.next(true);
        }
        return !response.expired;
      }),
      catchError(() => {
        this.clearToken();
        this.sessionExpired$.next(true);
        return new BehaviorSubject(false).asObservable();
      })
    );
  }

  // Forgot Password
  forgotPassword(email: string): Observable<any> {
    return this.http.post(`${this.apiUrl}/forgot-password`, { email }).pipe(
      catchError((error) => {
        return throwError(() => new Error(error.error?.message || 'An unexpected error occurred.'));
      })
    );
  }
}

File: Flutter/lib/services/auth_service.dart
import 'dart:convert';
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';

class AuthService extends ChangeNotifier {
  static final String _apiUrl = 'http://localhost:8000/api';

  // 'http://10.0.2.2:8000/api' Android emulator and make sure to run "php artisan serve --host=0.0.0.0 --port=8000"
  // 'http://localhost:8000/api' Web or other platforms

  bool _sessionExpired = false;

  bool get sessionExpired => _sessionExpired;

  Future<Map<String, dynamic>> register(Map<String, String> data) async {
    final response = await http.post(
      Uri.parse('$_apiUrl/register'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode(data),
    );
    return _handleResponse(response);
  }

  Future<Map<String, dynamic>> login(Map<String, String> data) async {
    final response = await http.post(
      Uri.parse('$_apiUrl/login'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode(data),
    );
    final result = _handleResponse(response);
    if (result['token'] != null) {
      await setToken(result['token']);
    }
    return result;
  }

  Future<void> deleteAccount(String password) async {
    final token = await getToken();
    final response = await http.post(
      Uri.parse('$_apiUrl/delete-account'),
      headers: {
        'Authorization': 'Bearer $token',
        'Content-Type': 'application/json',
      },
      body: jsonEncode({'password': password}),
    );
    _handleResponse(response);
    await clearToken();
  }

  Future<void> updateProfile(String name, String email) async {
    final token = await getToken();
    final response = await http.put(
      Uri.parse('$_apiUrl/update-profile'),
      headers: {
        'Authorization': 'Bearer $token',
        'Content-Type': 'application/json',
      },
      body: jsonEncode({'name': name, 'email': email}),
    );
    _handleResponse(response);
  }

  Future<Map<String, dynamic>> getProfile() async {
    final token = await getToken();
    final response = await http.get(
      Uri.parse('$_apiUrl/profile'),
      headers: {'Authorization': 'Bearer $token'},
    );
    return _handleResponse(response);
  }

  Future<void> updatePassword(
    String currentPassword,
    String newPassword,
  ) async {
    final token = await getToken();
    final response = await http.put(
      Uri.parse('$_apiUrl/update-password'),
      headers: {
        'Authorization': 'Bearer $token',
        'Content-Type': 'application/json',
      },
      body: jsonEncode({
        'current_password': currentPassword,
        'new_password': newPassword,
      }),
    );
    _handleResponse(response);
  }

  Future<void> forgotPassword(String email) async {
    final response = await http.post(
      Uri.parse('$_apiUrl/forgot-password'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({'email': email}),
    );
    _handleResponse(response);
  }

  Future<bool> checkTokenExpiry() async {
    final token = await getToken();
    if (token == null) {
      await clearToken();
      return false;
    }
    try {
      final response = await http.post(
        Uri.parse('$_apiUrl/check-token-expiry'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({'token': token}),
      );
      final data = jsonDecode(response.body);
      if (data['expired'] == true) {
        await clearToken();
        _sessionExpired = true;
        notifyListeners();
        return false;
      }
      return true;
    } catch (e) {
      await clearToken();
      _sessionExpired = true;
      notifyListeners();
      return false;
    }
  }

  Future<void> setToken(String token) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString('token', token);
    notifyListeners();
  }

  Future<String?> getToken() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString('token');
  }

  Future<void> clearToken() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove('token');
    notifyListeners();
  }

  Map<String, dynamic> _handleResponse(http.Response response) {
    if (response.statusCode >= 200 && response.statusCode < 300) {
      return jsonDecode(response.body);
    } else {
      throw Exception(response.body);
    }
  }
}
